Implement native wxTextCtrl spell checking V2.0

iwbnwif opened the pull request:
This is an update to the original Pull Request at #742 which in turn was based on wxTrac ticket https://trac.wxwidgets.org/ticket/17544.

The comments made against the original commit have been addressed in this new PR, but were answered in the original conversation for traceability. Comments can now be continued in this PR and referenced to the original PR #742 if necessary.

iwbnwif committed:
Add support for using native spell checking in wxTextCtrl

Add wxTextCtrl::EnableSpellCheck() and implement it for wxMSW and
wxGTK3.

Closes #17544.

iwbnwif commented:
Please note that only the input build files have been included in the first commit. Files that are automatically generated (e.g., by `autoconf`, `bakefile_gen` and `update-setup-h`) have not been included.

vadz commented:
Thanks for the update!

> Please note that only the input build files have been included in the first commit. Files that are automatically generated (e.g., by `autoconf`, `bakefile_gen` and `update-setup-h`) have not been included.

Unfortunately this means that the CI builds will fail. And, rather ironically, this also fails the spell checking test.

vadz commented on include/wx/msw/textctrl.h:
```diff
@@ -108,6 +108,13 @@ class WXDLLIMPEXP_CORE wxTextCtrl : public wxTextCtrlBase
 
+#if wxUSE_RICHEDIT && wxUSE_SPELLCHECK
+    // Use native spelling and grammer checking functions.
+    // This is only available on wxTE_RICH2 controls.
```
```suggestion
    // This is only available in wxTE_RICH2 controls.
```
(or maybe "for"?)

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -1017,6 +1023,52 @@ void wxTextCtrl::GTKSetJustification()
+}
+
+bool wxTextCtrl::IsSpellCheckEnabled() const
+{
```
It looks like this ought to check for `IsMultiLine()` too -- and maybe just return `false`? As it is, we're probably going to crash using null `textview` if this is called for a single line control.

iwbnwif replied:
I have added a guard for both IsMultiLine() and textview == NULL.

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -1017,6 +1023,52 @@ void wxTextCtrl::GTKSetJustification()
+    else
+    {
+        if ( spell )
+            gtk_spell_checker_detach(spell);
```
Shouldn't we destroy (i.e. unref) it here?

iwbnwif replied:
I don't think that it is necessary according to this page: http://gtkspell.sourceforge.net/doc/gtkspell3/tutorial-advanced.html

It looks like g_object_unref() is connected to the destroy signal and it is necessary to explicitly take another reference if you need to keep it?

iwbnwif commented:
> And, rather ironically, this also fails the spell checking test.
Yes, that made me laugh!

I will commit the generated build files now. I was expecting the CI scripts to run the tools.

vadz reviewed (commented):
Thanks again, but I think changes to `in{clude,terface}/wx/textctrl.h` got accidentally left out, could you please add them?

We also should ensure that the package containing `gtkspell.h` is installed during the CI builds to check compilation of this code.

Finally, it looks like it should be simple to implement this for macOS too as we already have `MacCheckSpelling()` there.

iwbnwif commented:
Any suggestions as to why `build/update-setup-h` is not correctly generating the platform header files properly are very welcome. I have barely touched `include/wx/setup_inc.h`. Could it be related to MSW vs Linux line endings (CR/CRLF)?

vadz commented:
> Any suggestions as to why `build/update-setup-h` is not correctly generating the platform header files are very welcome. I have barely touched `include/wx/setup_inc.h`. Could it be related to MSW vs Linux line endings (CR/CRLF)?

Yes, probably, because I can't reproduce this. Please make sure all your files use Unix EOLs (LF only), e.g. do **not** use Git auto CRLF translation if you're under MSW.

vadz commented on configure.in:
```diff
@@ -5761,6 +5762,37 @@ if test "$wxUSE_SECRETSTORE" = "yes"; then
+if test "$wxUSE_SPELLCHECK" = "yes"; then
+    dnl The required APIs are always available under MSW and OS X but we must
+    dnl have GNOME gtkspell under Unix to be able to compile this feature.
+    if test "$wxUSE_MSW" != "1" -a "$wxUSE_OSX_COCOA" != 1; then
+        if test "$wxGTK_VERSION" != "3"; then
```
I'm not sure if this does the right thing for the other ports, i.e. non-{GTK,MSW,OSX} ones. I think it could be better to write it as
```autoconf
if test "$wxUSE_MSW" = 1 -o test "$wxUSE_OSX_COCOA" = 1; then
    dnl Always available under these systems
    has_spellcheck_support=yes
elif test "$wxUSE_GTK" = 1; then
    PKG_CHECK_MODULES(....)
else
    dnl Not implemented in the other ports yet.
    has_spellcheck_support=no
fi

if test "$has_spellcheck_support" = yes; then
    AC_DEFINE(wxUSE_SPELLCHECK)
fi
```

vadz commented on include/wx/textctrl.h:
```diff
@@ -756,6 +756,12 @@ class WXDLLIMPEXP_CORE wxTextCtrlBase : public wxControl,
 
+#if wxUSE_SPELLCHECK
+    // Use native spelling and grammar checking functions.
+    virtual bool EnableSpellCheck(bool enable = true) { (void)enable; return false; }
```
Just for style consistency:
```suggestion
    virtual bool EnableSpellCheck(bool WXUNUSED(enable) = true) { return false; }
```

iwbnwif replied:
Sorry, I forgot that WXUNUSED went in the declaration - I was trying it in the implementation which of course didn't work!

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -1017,6 +1023,55 @@ void wxTextCtrl::GTKSetJustification()
+{
+    GtkTextView *textview = GTK_TEXT_VIEW(m_text);
+
+    if ( !IsMultiLine() || textview == NULL)
```
Extremely minor, but
```suggestion
    if ( !IsMultiLine() || textview == NULL )
```

iwbnwif replied:
> Thanks again, but I think changes to `in{clude,terface}/wx/textctrl.h` got accidentally left out, could you please add them?

Sorry, they are added now.

> We also should ensure that the package containing `gtkspell.h` is installed during the CI builds to check compilation of this code.

Is this not checked by PKG_CHECK_MODULES(GTKSPELL, [gtkspell3-3.0], ...? I will make the changes to configure.in that you mentioned above - but is there another step needed?

> Finally, it looks like it should be simple to implement this for macOS too as we already have `MacCheckSpelling()` there.

Yes, it looks almost trivial except for the IsSpellCheckEnabled() part. This (probably) requires a change to cocoa/textctrl.mm file - again quite simple, but I have no means of testing other than by the CI tools. Therefore, I would prefer to do this in a separate branch once I have finished the wxMSW and wxGTK stuff.

vadz replied:
> Is this not checked by PKG_CHECK_MODULES(GTKSPELL, [gtkspell3-3.0], ...? I will make the changes to configure.in that you mentioned above - but is there another step needed?

I was thinking of installing the required packages in `.github/workflows/ci.yml` as currently they're not available in the CI builds, as can be seen from
```
configure: WARNING: gtkspell3-3.0 not found, spellchecking won't be available
```
message in the build logs.

As for Mac part, I think I can do it, so please don't worry about this.

TIA!

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -829,6 +835,39 @@ WXDWORD wxTextCtrl::MSWGetStyle(long style, WXDWORD *exstyle) const
+            "Unable to enable spell check on control "
+            "which does not have wxTE_RICH2 style");
+
+    LPARAM editStyle = SES_USECTF | SES_CTFALLOWEMBED
```
Looks like we need `SES_XXX` definitions for MinGW (not -w64 variety) too, could you please add them above together with `IMF_SPELLCHECKING `?

iwbnwif replied:
Yes, I have been looking at doing this. I was thinking that Appveyor must have a pretty old version of MinGW as they have been there since 2014, but that is MinGW -w64 as you say. 

I will add them alongside IMF_SPELLCHECKING - wasn't sure if this was the best place or in missing.h.

vadz replied:
I guess we're going to finally drop support for classic MinGW after 3.2.0, together with the other old compilers, but not yet...

As for where to define it, the heuristic is that if it's likely to be needed by multiple files, then it should go into `missing.h`, otherwise be defined just in the only file using it. Here it seems unlikely that we should need them elsewhere, so I'd keep things local.

TIA!

vadz reviewed (commented):
Looks good, thanks, the only thing that needs to be done before merging is to fix the build with MinGW by defining the missing constants.

After looking at this once again, I wonder if it we should pass `int flags = Text_SpellCheck` rather than `bool enable = true` to `EnableSpellChecking()`, in case we want to provide more options for this function later (e.g. enabling spell but _not_ grammar checks)?

vadz commented:
Thanks for the update! I'll try to add Mac support if I have time but will merge this anyhow even if I don't.

Unless you think it's worth changing `bool enable` to `int flags`?

vadz commented:
P.S. At least Mac has both "spell checking" and "grammar checking", see https://developer.apple.com/documentation/appkit/nstextview?language=objc, so I think it might well be worth to have different flags for enabling them. OTOH `GtkSpellChecker` doesn't have anything like this, but does have the possibility to specify the language, which might also be supported under MSW if we decided to use `ISpellCheckerFactory` directly. So perhaps we need not just `flags` per se, but something like `wxSpellCheckOptions` object, with the default one enabling all checks for the current language? Of course, we don't need to implement the full flexibility right now, but I just think that we may regret restricting the parameter to just a `bool` here later.

iwbnwif commented:
> Of course, we don't need to implement the full flexibility right now, but I just think that we may regret restricting the parameter to just a bool here later.

Yes, no problem. I saw that OSX and MSW have more options. My inclination would be something like:

EnableSpellCheck(bool enable, int options = wxTE_SPELL_OPTION_ALL, lang="")

Other options would be wxTE_SPELL_OPTION_SPELL and wxTE_SPELL_OPTION_GRAMMAR. Then these would have to be OR'able where options are not mutually exclusive. lang = "" means whatever is set for the app, others would be ISO 639-1 "fr", "fr-BE", "en-US" etc.

A wxSpellCheckOptions object is also okay - I can also see that the ability to set (and change) options globally or individually would be of value. Can it be coded with a couple of options + language now and then others added later? Any effect on ABI?

I honestly can't think of anything other than selecting spellcheck, grammar check and the language that are of any practical use. But granted there are others, such as disabling context menu, personal dictionaries etc.

vadz commented:
> A wxSpellCheckOptions object is also okay - I can also see that the ability to set (and change) options globally or individually would be of value. Can it be coded with a couple of options + language now and then others added later? Any effect on ABI?

An object is mostly nice because it allows to add more things later without changing the API (you'd think that you can always add more optional arguments later, and you'd be mostly right, but sometimes you run into ambiguities with different overloads and other times you end up with a function taking 9 different parameters in 10 years, which is not great neither), but we could indeed future-proof it from the ABI point of view by adding `wxString m_lang` and `int m_options` fields to it right now, even if we don't use them yet.

iwbnwif commented:
I had a quick look at the current options available on the different platforms and came up with the following:

```
#define wxTE_SPELL_OPTION_ALL_CHECKS 0
#define wxTE_SPELL_OPTION_SPELL 1
#define wxTE_SPELL_OPTION_GRAMMAR 2
#define wxTE_SPELL_OPTION_AUTOCORRECT 4
#define wxTE_SPELL_OPTION_NO_CONTEXT_MENU 8
#define wxTE_SPELL_OPTION_NO_IGNORE_WORD 16
#define wxTE_SPELL_OPTION_NO_PERSONAL_DICTIONARY 32
```
As you say, I am sure there are others, but I think that we can cover the current platform options with a single int. So, I am still inclined in that direction, but not strongly against an object either.

vadz commented:
Yes, options could be covered by a single int (this was my initial `int flags` proposal), but if we want to also allow (optionally) setting the language, we need either another argument or an object. 2 arguments wouldn't be too bad, of course, but if we need to add something else later (e.g. custom dictionary path), it would start becoming unwieldy, while with an object we can have something akin to "named" arguments using `wxSizerFlags`-like API, i.e. `EnableSpellCheck(wxSpellCheckOptions("fr").EnableGrammarChecks())`.

iwbnwif commented:
That's strange - the most recent message only just showed up, even though it is time stamped 3 hours ago.

Anyway, with regard to the options - I can only say that either way is fine by me!

iwbnwif commented on include/wx/textctrl.h:
```diff
@@ -127,19 +127,42 @@ enum wxTextCtrlHitTestResult
+    bool m_EnableGrammerCheck;
 };
 
+extern WXDLLIMPEXP_DATA_CORE(const wxTextProofOptions) wxDefaultTextProofOptions;
```
Please check that this declaration is correct.

iwbnwif replied:
On second thoughts, do we even need this? 

I think it would be okay just to pass the default `wxTextProofOptions` constructor as the default parameter, e.g., 

```
    virtual bool EnableProofCheck(bool WXUNUSED(enable) = true,
                                  const wxTextProofOptions& WXUNUSED(options) =
                                  wxTextProofOptions()) wxOVERRIDE;
```

vadz replied:
Yes, it is correct and yes, I also agree that we don't need it, using `= wxTextProofOptions()` is also perfectly fine.

vadz commented on include/wx/textctrl.h:
```diff
@@ -126,6 +126,43 @@ enum wxTextCtrlHitTestResult
+        return *this;
+    }
+
+    wxTextProofOptions& GrammerCheck(bool enable = true)
```
Sorry, a spelling fix (sic) is needed here:
```suggestion
    wxTextProofOptions& GrammarCheck(bool enable = true)
```

iwbnwif replied:
:$

vadz commented on include/wx/textctrl.h:
```diff
@@ -126,6 +126,43 @@ enum wxTextCtrlHitTestResult
+private:
+    wxString m_lang;
+    bool m_EnableSpellCheck;
+    bool m_EnableGrammerCheck;
```
```suggestion
    bool m_EnableGrammarCheck;
```

vadz commented on interface/wx/textctrl.h:
```diff
@@ -74,6 +74,22 @@ enum wxTextAttrAlignment
     wxTEXT_ALIGNMENT_JUSTIFIED
 };
 
+/**
```
This is out of date/not used?

vadz commented on interface/wx/textctrl.h:
```diff
@@ -1336,6 +1351,38 @@ class wxTextCtrl : public wxControl,
+        the wxTE_RICH2 style set. wxGTK3 requires that the control has the
+        wxTE_MULTILINE style.
+
+        @param enable
```
This needs to be updated to correspond to the actual API.

iwbnwif replied:
Yes, will do that now.

vadz commented on src/gtk/textctrl.cpp:
```diff
@@ -1017,6 +1023,57 @@ void wxTextCtrl::GTKSetJustification()
+        wxString lang = options.GetLang();
+
+        if ( !gtk_spell_checker_set_language(spell,
+                                    lang.IsEmpty() ? NULL : (const gchar *)lang.c_str(),
```
```suggestion
                                    lang.empty() ? NULL : (const gchar *)lang.utf8_str(),
```
First change is just a matter of taste (I think standard-like methods should be preferred), but the second one might be more important if we ever have languages using non-ASCII characters (currently it can't happen AFAIK but why not support them if it ever does, as long as it doesn't cost anything).

iwbnwif replied:
I will make these updates on my local copy to avoid a merge (I've implemented `=  wxTextProofOptions()` locally).

vadz commented on src/msw/textctrl.cpp:
```diff
@@ -812,6 +832,39 @@ WXDWORD wxTextCtrl::MSWGetStyle(long style, WXDWORD *exstyle) const
 
+#if wxUSE_RICHEDIT && wxUSE_SPELLCHECK
+
+bool wxTextCtrl::EnableProofCheck(bool enable, const wxTextProofOptions& options)
```
AFAICS this is not used here currently:
```suggestion
bool wxTextCtrl::EnableProofCheck(bool enable, const wxTextProofOptions& WXUNUSED(options))
```

vadz reviewed (commented):
Thanks, looks good to me!

vadz commented:
Please let me know if this is ready to be merged now -- I'll check if I can quickly extend support to Mac and merge it then. TIA!

iwbnwif commented:
> Please let me know if this is ready to be merged now -- I'll check if I can quickly extend support to Mac and merge it then. TIA!

I am just waiting for the CI to run, but assuming no failures then it should be fine to merge. Good luck!

vadz committed:
Verify that wxUSE_SPELLCHECK is always defined

Add sanity check for this symbol, as it's done for all the other
wxUSE_XXXs.

vadz committed:
Always set wxUSE_SPELLCHECK to 1 by default for simplicity

It does no real harm to set it to 1 and this avoids the need for the
platform checks in setup.h files, simplifying things (although it does
require checking for GTK 3 in the code compiled for both GTK 2 and 3, as
gtkspell library can only be used with GTK 3).

vadz committed:
Reduce configure warnings about missing spell checking support

It's not really useful to warn that spell checking is not available for
the ports for which it's not implemented anyhow, as there is nothing
that the person running configure can do about it anyhow.

Still warn about missing gtkspell library when using GTK 3, however, as
it may be useful to know that it's not available and could be installed
in order to activate spell checking support.

vadz committed:
Slightly reformat EnableProofCheck() declarations

Keep the entire wxTextProofOptions argument on the same line.

Remove unnecessary WXUNUSED() in the declarations of the function in
wxMSW/wxGTK headers.

No real changes.

vadz committed:
Mark wxTextProofOptions as being new in 3.1.6 in the docs too

vadz committed:
Implement minimal support for spell checking in wxOSX too

Reuse the existing CheckSpelling() function.

Also deprecate wxTextCtrl::MacCheckSpelling() in favour of the new
portable EnableProofCheck().

vadz commented:
I've wired this up to the existing Mac function and it works fine, but while doing it, I've realized that I don't really like having  both `bool enable` and `wxTextProofOptions` arguments, I think having just one should be enough. I.e. in the unlikely case we want to disable spell checking (to be honest, I don't see any realistic scenario in which we'd need to do it), we'd just use `EnableProofCheck(wxTextProofOptions::DisableAll())`.

Do you have any objections to this? If not, I'm going to do it like this.

TIA!

iwbnwif commented:
> I.e. in the unlikely case we want to disable spell checking (to be honest, I don't see any realistic scenario in which we'd need to do it), we'd just use EnableProofCheck(wxTextProofOptions::DisableAll()).

I need to be able to enable and disable spell checking on a particular control at run time. The controls are used to enter data into a database. Normally, a description field is natural language in which case spell checking is useful. In some use cases though, the field is used for entering non-natural language text and some people find it annoying to have red lines all the time.

> Do you have any objections to this?

No, it's fine. I can always `#define EnableProofCheck(wxTextProofOptions::DisableAll()) DisableProofCheck()` :)

vadz committed:
Don't use DLL export declaration with wxTextProofOptions

This class doesn't have any non-inline methods, so applying DLL export
declaration to it is at best useless and might even be harmful, so don't
do it.

vadz committed:
Use wxTextProofOptions itself to enable or disable the checks

Remove a separate "bool enable" argument of EnableProofCheck() and use
wxTextProofOptions::IsSpellCheckingEnabled() to decide whether the
checks should be enabled or disabled.

Also remove wxTextProofOptions ctor and provide named static factory
functions for creating the objects of this class with clearly defined
meaning.

vadz committed:
Don't show message box in text sample unnecessarily

This is going to be just annoying on the platforms where spell checking
is not supported, so show the message inside the text control itself
rather than popping up a message box.

Also put a misspelled word in the control from the very beginning to
show that it is highlighted.

vadz committed:
Implement support for grammar checking in wxOSX

Use native support for grammar checking in NSTextView.

As we're not passing wxTextProofOptions to the lower level function,
this functionality now depends on wxUSE_SPELLCHECK, meaning that even
the previously existing MacCheckSpelling() function is not defined any
more when wxUSE_SPELLCHECK is set to 0. This is not completely
backwards-compatible, but hopefully shouldn't be a problem in practice
and shouldn't break any existing applications which can't disable the
just added wxUSE_SPELLCHECK.

vadz committed:
Correct description of EnableProofCheck() requirements

This functionality is available since Windows 8 only, not 7.

vadz committed:
Remove stub spell checking functions from wxiOS wxTextCtrl

There doesn't seem to be any need to have these functions here as the
base class implementation already does exactly the same thing (i.e.
nothing at all).

vadz committed:
Replace IsProofCheckEnabled() with GetProofCheckOptions()

This is more consistent with EnableProofCheck() and allows to retrieve
the current state of grammar checking under macOS, which can be checked
by user and so can be useful to know.

vadz commented:
I've pushed my changes now. I've tested them under Mac and they seem to work fine there, so if there are no objections, I think this is ready for merging.

I think it would be probably better to squash your commits to avoid "correct typo" and "remove whitespace" commits that are not really useful, so I'm going to do if you don't mind before merging -- or please feel free to do it yourself, especially if you'd like to keep your changes in more than one commit (if I do the squashing, I'm going to squash all of them together).

Finally, we could add a trivial `wxTextCtrl::DisableProofCheck()` if you think it's worth it. IMO it isn't as it shouldn't be used that often, but please feel free to add it if you feel strongly about this.

And finally finally I think it would be also useful to have XRC support for this, but this can certainly wait for another day/PR.

iwbnwif commented:
> I think it would be probably better to squash your commits to avoid "correct typo" and "remove whitespace" commits that are not really useful, so I'm going to do if you don't mind before merging -- or please feel free to do it yourself, especially if you'd like to keep your changes in more than one commit (if I do the squashing, I'm going to squash all of them together).

Please do squash them,  I think this can be a single commit, including your changes.

I would normally do that myself locally and only push the squashed commits, but I haven't got my setup back to a state where I can properly test different builds and do whitespace checks etc. Therefore, the CI tests were a convenient [lazy!] way to do that.

> Finally, we could add a trivial wxTextCtrl::DisableProofCheck() if you think it's worth it. IMO it isn't as it shouldn't be used that often, but please feel free to add it if you feel strongly about this.

No, I don't think it is necessary at this point. My longer term idea is to be able to set proof check options on a top level window or wxStaticBox and for all editable controls in that window / group to inherit the settings if they have an appropriate {style (whispered)} setting. I think `DisableProofCheck()` may be more useful at that point to support a user option.

> And finally finally I think it would be also useful to have XRC support for this, but this can certainly wait for another day/PR.

Understood. We also need to see what is happening with `gspell `and GTK4 :)

iwbnwif commented on include/wx/osx/textctrl.h:
```diff
@@ -97,6 +97,16 @@ class WXDLLIMPEXP_CORE wxTextCtrl: public wxTextCtrlBase
 
+#if wxUSE_SPELLCHECK
+    // Use native spelling and grammar checking functions.
+    // This is only available in wxTE_RICH2 controls.
```
Is wxTE__RICH2 appropriate for OSX?

vadz replied:
Oops, no, thanks for noticing this.

iwbnwif commented on samples/text/text.cpp:
```diff
@@ -1209,7 +1209,7 @@ MyPanel::MyPanel( wxFrame *frame, int x, int y, int w, int h )
 
 #if wxUSE_SPELLCHECK
-    if ( !m_enter->EnableProofCheck(true, wxTextProofOptions("en_US").SpellCheck()) )
+    if ( !m_enter->EnableProofCheck(wxTextProofOptions::Default().GrammarCheck()) )
```
Is it worth adding a note to say that `.GrammarCheck()` is added as an example, but not currently supported?

iwbnwif commented on include/wx/textctrl.h:
```diff
@@ -126,6 +126,76 @@ enum wxTextCtrlHitTestResult
+    }
+
+    // And the corresponding accessors.
+    bool IsSpellCheckingEnabled() const { return m_EnableSpellCheck; }
```
Please can this be `IsSpellCheckEnabled()` to be consistent with the enabling function (`SpellCheck()`)

vadz replied:
OK, I'll rename all of them as suggested.

iwbnwif commented on include/wx/textctrl.h:
```diff
@@ -126,6 +126,76 @@ enum wxTextCtrlHitTestResult
+
+    // And the corresponding accessors.
+    bool IsSpellCheckingEnabled() const { return m_EnableSpellCheck; }
+    bool IsGrammarCheckingEnabled() const { return m_EnableGrammarCheck; }
```
Also `IsGrammarCheckEnabled()` please.

iwbnwif commented on include/wx/osx/cocoa/private/textimpl.h:
```diff
@@ -129,7 +129,7 @@ class wxNSTextViewControl : public wxNSTextBase
 #if wxUSE_SPELLCHECK
     virtual void CheckSpelling(const wxTextProofOptions& options) wxOVERRIDE;
-    virtual bool IsSpellingCheckEnabled() const wxOVERRIDE;
+    virtual wxTextProofOptions GetCheckingOptions() const wxOVERRIDE;
```
Can't this be `GetProofCheckOptions()` too or is it for backwards compatibility with an existing function?

vadz replied:
No, this is a private header, so there are no compatibility concerns here, but I find calling these private functions using the same names as the public ones confusing, so I wanted to use a different name. And this one is consistent with what Cocoa uses (the properties enabling different checks are called `xxxChecking`), so I'd rather leave it as is.

Again, this is invisible to wx users anyhow and can always be changed later if really needed.

iwbnwif commented on include/wx/osx/core/private.h:
```diff
@@ -741,7 +741,7 @@ public :
 #if wxUSE_SPELLCHECK
     virtual void CheckSpelling(const wxTextProofOptions& WXUNUSED(options)) { }
-    virtual bool IsSpellingCheckEnabled() const { return false; }
+    virtual wxTextProofOptions GetCheckingOptions() const;
```
Can't this be GetProofCheckOptions() too or is it for backwards compatibility with an existing function?

iwbnwif commented on include/wx/textctrl.h:
```diff
@@ -175,6 +175,11 @@ class wxTextProofOptions
 
+    bool AnyChecksEnabled() const
+    {
+        return IsSpellCheckingEnabled() || IsGrammarCheckingEnabled();
```
Suggest `return IsSpellCheckEnabled() || IsGrammarCheckEnabled();`

vadz commented:
I've rebased and force pushed the changes now and I think they should be good to merge.

Thanks again!

iwbnwif commented:
Great, thank you for filling all the gaps and tidying up. The good news is that this should close 2 PR and one Trac ticket!

vadz commented:
Returning to this once again, there seem to be some reasons to [prefer gspell](https://gitlab.gnome.org/GNOME/gspell/-/blob/master/docs/faq.md) when using wxGTK. OTOH this project [seems to be dead](https://gitlab.gnome.org/GNOME/gspell/-/blob/master/HACKING), so I'm not sure if it's a good idea to start using it now.

OTOH it's not like there are any [recent commits](https://sourceforge.net/p/gtkspell/gtkspell/commit_browser) to GtkSpell neither.

Does anybody know if gspell is still the recommended spell-checking library to use for GTK?

FWIW under Debian gspell is a dependency of the following packages:
- geary
- osmo
- latexila
- gtranslator
- balsa
- gnome-software
- gnome-recipes
- gnome-builder
- gedit
- libevview3-3

while gtkspell is a dependency of

- inkscape
- poedit (so I wonder if @vslavik has any plans to transition to gspell?)
- pan
- modem-manager-gui
- liblablgtk3-ocaml
- kildclient
- gnome-subtitles
- gnote
- almanah
- gitg
- geany-plugin-vc
- libevolution

so both seem to be used in several popular projects.

Anyhow, unless someone comes up and says that we definitely shouldn't be using gtkspell, I'm still going to merge this in its current state, as it looks like we ought to be able to switch to gspell without too much trouble later, but OTOH if we do decide to switch, it would be better to do it before 3.1.6.
